src/Agent.Worker/CommandHandler.cs

```diff
@@ -26,6 +26,7 @@ public override void Initialize(IHostContext hostContext)
             var extensionManager = hostContext.GetService<IExtensionManager>();
             foreach (var commandExt in extensionManager.GetExtensions<ICommandExtension>() ?? new List<ICommandExtension>())
             {
+                // TODO: I am seeing same command extension registed twice in diag trace, something goes wrong?
```

Comment:
It's because GetService uses the non-blocking pattern to create the cached instances. If two requests come in parallel, two objects may be created but one gets thrown away. Once an instance gets cached, it's not overwritten. It's an often recommended approach over the "double-check and lock" pattern.

The typical way it's done is:
```
if (_cache == null) {
  Interlocked.CompareExchange(location1: ref cache, value: new Foo(), comparand: null);
}

return _cache;
```

GetService is similar but uses TryAdd because it's dealing with a dictionary of many different types.